t/proxyunits.t

```diff
@@ -120,14 +126,18 @@ sub proxy_test {
         }
     }
 
-    # verify proxy received response
-    if (scalar @{$ps_recv}) {
-        foreach my $recv (@{$ps_recv}) {
-            is(scalar <$ps>, $recv, "ps returned expected response.");
+    if (defined $ps_recv) {
+        if (scalar @{$ps_recv}) {
+            foreach my $recv (@{$ps_recv}) {
+                is(scalar <$ps>, $recv, "ps returned expected response.");
+            }
+            # makes sure nothing remains in $ps.
+            check_version($ps);
+        } else {
+            # when $ps_recv is empty, make sure it is not readable.
+            my @readable = $pss->can_read(0.1);
```

Comment:
not 100% sure I follow what's going on here. can it simply do the check_version either way? Otherwise it's going to add a 0.1s delay every time a test like this is run.

Reply:
I am debating on this too, the behavior I am trying to support if

First `proxy_test`, some backends responded, and expect $ps should receive none.
Second `proxy_test`, other backends responded, $ps received response.

More generically, it allows inserting actions among multiple `proxy_test` calls for complex scenarios.

> Otherwise it's going to add a 0.1s delay every time a test like this is run.

Yes, but unless we explicitly do `ps_recv => [],` which only happens in a single scenario in the quiet flag test.

Reply:
I'm not complaining I'm just not (yet?) seeing how it's functionally different.

either you check if the socket is readable (which only happens if something was sent)
or you put something in the after the test runs and the response (a version check). if you get that response, there was nothing else to be read.

they should be functionally equivalent (if the socket became readable that's because there was something in there), but in the second case it can execute immediately instead of after 100ms.

unless I'm still misunderstanding?

Reply:
```
    if (defined $ps_recv) {
        if (scalar @{$ps_recv}) {
            foreach my $recv (@{$ps_recv}) {
                is(scalar <$ps>, $recv, "ps returned expected response.");
            }
            # makes sure nothing remains in $ps.
            check_version($ps);
        } else {
            # when $ps_recv is empty, make sure it is not readable.
            my @readable = $pss->can_read(0.1);
            is(scalar @readable, 0, "ps is expected to be non-readable");
+++     check_version($ps); 
        }
    }
```


This test will fail with added check version
```
    subtest 'await(r, p, 1): send [e, m, h], recv 3 responses and 1 reconn' => sub {
        my $key = "/awaitone/a";
        my $ps_send = "get $key\r\n";
        my $be_hit = "VALUE $key 0 2\r\nok\r\nEND\r\n";
        my $be_miss = "END\r\n";
        my $be_error = "ERROR backend failure\r\n";
        proxy_test(
            ps_send => $ps_send,
            be_recv => {0 => [$ps_send], 1 => [$ps_send], 2 => [$ps_send]},
            be_send => {0 => [$be_error], 1 => [$be_miss]},
            ps_recv => [],
        );
        proxy_test(
            be_send => {2 => [$be_hit]},
            ps_recv => ["VALUE $key 0 1\r\n", "3\r\n", "END\r\n"],
        );
        # reconnect due to ERROR
        $mbe[0] = accept_backend($mocksrvs[0]);
    };
```

failure is 
```
    #   Failed test 'version received'
    #   at t/proxyunits.t line 74.
    #                   'VALUE /awaitone/a 0 1
    # '
    #     doesn't match '(?^:VERSION )'

    #   Failed test 'ps returned expected response.'
    #   at t/proxyunits.t line 132.
    #          got: '3
    # '
    #     expected: 'VALUE /awaitone/a 0 1
    # '
```

I assume this is because `VERSION` unblocked await(), but please correct me.

Reply:
works without added check_version()

Reply:
okay so I guess when you send the version it's going to block until there's a timeout, then you'll get some response from the await as it will complete. which is weird since it looks like you're not getting an error or a miss, you're getting a VALUE. So maybe something's actually broken somewhere? or elsewhere in the test structure?

Reply:
await() is unblocked as confirmed by print from Lua after await()

Reply:
yeah; but your example is showing a `VALUE` response during the `check_version` call, but you're not sending that from the backend earlier. So where is that coming from? or is the example copy/pasted wrong?

Reply:
To be clear I'm saying "I smell a bug", unless the example you copy/pasted was wrong.

Reply:
yeah, I am seeing an unexpected timeout event, which marked the third response as error, but I am not yet clear why timeout can happen.

Reply:
I'm not talking about the timeout; I'm talking about the string you're getting back that you didn't send. The timeout makes sense.

Reply:
The timeout will happen when you send a version command because the test will block until it gets _any_ response, so the backend waiting for a response to the await will eventually time out and error, letting the await complete, then send what it had upstream.

but in your example you send an error and a miss, but get a `VALUE` back during the version test. that should be a bug somewhere?

Reply:
> I'm not talking about the timeout; I'm talking about the string you're getting back that you didn't send. 

That is hard coded in Lua, when await() is unblocked. So timeout unblocked the await() and return that string

Reply:
Sorry you might be right, I need some more time to sort this out.

Reply:
Derp. I'm an idiot. But now that I'm looking at the right place it's more confusing. but I think I get it:

- awaitone/a is waiting for "1" response. The default mode is GOOD (looking for hits, ideally)
- in your test you send an error and a miss, then it dies and times out because of the weird modification I asked for because I'm dumb.
- await returns all of its responses so far, which ends up being 1 (the miss, I'm guessing).

So either A) That error should've also been turned into a response? I think not but can't remember why it doesn't offhand. IIRC backend failures are harder errors than an error message type.
or B) awaitone should be counting the number of `ok()` or good or hits instead of just the total number. Since it looks like we can accidentally write buggy tests with it without being more clear.

Reply:
OK, I think I got what you meant.

So if we sent await() and version to proxy, the return sequence is await() and version, meaning version response is always after await().

then timeout makes sense, like you explained.

And to your question, again the VALUE string is hard-coded in Lua, that always returns after await(), make sense?

Reply:
Circling back to the original problem; I think once we're doing magical state across multiple calls to `proxy_test` via special arguments that change behavior it no longer meets my threshold for simplicity. If that's okay?

The readable wait you added can stay though. now that I actually understand why it's there.

Reply:
OK I will refrain from using multiple `proxy_test` for complex tests.

Sure we can verify the count of ok and none oks